src/transformStateWithClones.js

```diff
@@ -7,7 +7,34 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const versions = [];
+  const currentState = Object.assign({}, state);
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        Object.assign(currentState, action.extraData);
+        break;
+
+      case 'removeProperties':
+        for (const prop of action.keysToRemove) {
+          delete currentState[prop];
+        }
+        break;
+
+      case 'clear':
+        for (const prop in currentState) {
```

Comment:
When the 'clear' action is encountered, the currentState is cleared by deleting each property. A more efficient approach would be to assign an empty object to currentState.